### PR TITLE
Tickets/dm 36418

### PR DESCRIPTION
**Pull request by fjammes**

### CODE REVIEW

File: codeformat.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+DIR=$(cd "$(dirname "$0")"; pwd -P)
+
+black "$DIR"
+flake8 "$DIR"
```
Comment (andy-slac) [inline codeformat.sh:8]: In middleware people use pre-commit hooks which work reasonably OK and do not require separate script which you can forget to run. We use `pre-commit` package: https://pre-commit.com/, and you can check an example of configuration here: https://github.com/lsst/daf_butler/blob/main/.pre-commit-config.yaml.

File: doc/conf.py [hunk 2]
```diff
@@ -58,10 +58,7 @@
 
 # Add any URL patterns to ignore (e.g. for private sites, or sites that
 # are frequently down).
-linkcheck_ignore = [
-    r"^https://jira.lsstcorp.org/browse/",
-    r"^https://dev.lsstcorp.org/trac"
-]
+linkcheck_ignore = [r"^https://jira.lsstcorp.org/browse/", r"^https://dev.lsstcorp.org/trac"]
```
Comment (andy-slac) [inline doc/conf.py]: We should not have any links to Trac at this point?

File: rootfs/ingest/python/qserv/contribution.py [hunk 3]
```diff
@@ -91,8 +92,9 @@ def __init__(
                 self.ext = ext
 
         if len(self.ext) == 0:
-            raise IngestError("Unsupported data format for regular table "
-                              f"only {metadata.EXT_LIST} are supported")
+            raise IngestError(
+                "Unsupported data format for regular table " f"only {metadata.EXT_LIST} are supported"
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Should merge two strings together (and move `f` to the front).

File: rootfs/ingest/python/qserv/contribution.py [hunk 4]
```diff
@@ -128,17 +130,15 @@ def _build_payload(self, transaction_id: int) -> dict:
         return payload
 
     def start_async(self, transaction_id: int) -> None:
-        """Start an asynchronous ingest query for a chunk contribution
-           Raise an exception if the query fails after a fixed number of attempts
-           (see MAX_RETRY_ATTEMPTS constant)
+        """Start an asynchronous ingest query for a chunk contribution Raise an
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Needs period after "contribution".

File: rootfs/ingest/python/qserv/contribution.py [hunk 5]
```diff
@@ -164,23 +164,26 @@ def start_async(self, transaction_id: int) -> None:
...
     def monitor(self) -> bool:
-        """Monitor an asynchronous ingest query for a chunk contribution
-           States of an ingest query are listed here:
-           https://confluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol
-           Propagate an exception if the query fails after a fixed number of attempts
+        """Monitor an asynchronous ingest query for a chunk contribution States
+        of an ingest query are listed here:
+        https://confluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol
+        Propagate an exception
+        if the query fails after a fixed number of attempts.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Line break probably not needed (or maybe later in the sentence).

File: rootfs/ingest/python/qserv/contribution.py [hunk 6]
```diff
@@ -164,23 +164,26 @@ def start_async(self, transaction_id: int) -> None:
...
-                Raised for unmanaged contribution state or non-retriable errors from R-I system
+                Raised for unmanaged contribution state or non-retriable errors
+                from R-I system
 
 
         Returns
         -------
             bool: True if contribution has been successfully ingested
                   False if contribution is being ingested
+
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py:185]: This empty line is not needed.

File: rootfs/ingest/python/qserv/contribution.py [hunk 7]
```diff
@@ -164,23 +164,26 @@ def start_async(self, transaction_id: int) -> None:
...
+        if the query fails after a fixed number of attempts.
 
         Raises
         ------
             IngestError
                 Raised in case of error during contribution ingest
             ReplicationControllerError
-                Raised for unmanaged contribution state or non-retriable errors from R-I system
+                Raised for unmanaged contribution state or non-retriable errors
+                from R-I system
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Just one empty line is enough

File: rootfs/ingest/python/qserv/contribution.py [hunk 8]
```diff
@@ -218,14 +223,14 @@ def monitor(self) -> bool:
...
+                    f"Contribution {self} is in status {contrib_monitor.status} "
+                    f"with error: {contrib_monitor.error}, "
+                    f"system error: {contrib_monitor.system_error}, "
+                    f"http error: {contrib_monitor.http_error}"
+                )
                 if noretry_errmsg:
-                    raise IngestError(
-                        msg + f" {noretry_errmsg}"
-                    )
+                    raise IngestError(msg + f" {noretry_errmsg}")
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/contribution.py]: Maybe

    f"{msg} {noretry_errmsg}"

to keep things uniform?

File: rootfs/ingest/python/qserv/http.py [hunk 9]
```diff
@@ -153,10 +160,30 @@ def _authenticate(self, auth_path: Optional[str]) -> str:
...
+    def get(
+        self, url: str, payload: Dict[str, Any] = dict(), auth: bool = True, timeout: Optional[int] = None
+    ) -> Dict:
+        """Send a GET query to an http(s) URL.
+
+        Parameters
+        ----------
+        url : `str`
+            Http(s) URL
+        payload : `Optional[Dict[str, Any]`
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: Need second closing bracket. 

Complex type formatting is done differently, You need something like:

    payload : `Dict` [`str`, `Any`], optional

see   https://developer.lsst.io/python/numpydoc.html#sequence-types

File: rootfs/ingest/python/qserv/http.py [hunk 10]
```diff
@@ -153,10 +160,30 @@ def _authenticate(self, auth_path: Optional[str]) -> str:
...
+            JSON payload, Defaults to dict().
+        auth : `Optional[bool]`
+            Perform HTTP authentication. Defaults to True.
+        timeout : `Optional[int]`
+            Query time-out. Defaults to None.
+
+        Raises:
+            ReplicationControllerError: JSON response contain an error code
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: Raises
    ------
    ReplicationControllerError
        Raised if JSON response contain an error code

File: rootfs/ingest/python/qserv/http.py [hunk 11]
```diff
@@ -153,10 +160,30 @@ def _authenticate(self, auth_path: Optional[str]) -> str:
...
+
+        Returns:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: Returns
    -------

File: rootfs/ingest/python/qserv/http.py [hunk 12]
```diff
@@ -168,8 +195,30 @@ def get(self, url: str,
...
+        response_json : `dict`
+            JSON response
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: Same here, please pay attention to proper docstring formatting. Please check all changes, I'm going to ignore these below.

File: rootfs/ingest/python/qserv/http.py [hunk 13]
```diff
@@ -168,8 +195,30 @@ def get(self, url: str,
         _LOG.debug("GET: success")
         return response_json
 
-    def post(self, url: str, payload: Dict[str, Any] = dict(),
-             auth: bool = True, timeout: int = None) -> Dict:
+    def post(
+        self, url: str, payload: Dict[str, Any] = dict(), auth: bool = True, timeout: int = None
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/http.py]: It is not OK to use `dict()` as a default argument value: https://developer.lsst.io/python/style.html#a-mutable-object-must-not-be-used-as-a-keyword-argument-default
Please check all new methods below.

File: rootfs/ingest/python/qserv/ingest.py [hunk 14]
```diff
@@ -76,6 +75,7 @@ class Ingester:
             Replication controller URL
         queue_manager: `QueueManager`
             Manager to access contribution queue
+
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: No need for thes empty line

File: rootfs/ingest/python/qserv/ingest.py [hunk 15]
```diff
@@ -110,39 +110,36 @@ def check_supertransactions_success(self) -> None:
...
 
         Parameters
         ----------
         replication_config: `util.ReplicationConfig`
-            Configuration parameters for the database inside replication/ingest system
+            Configuration parameters for the database inside
+            replication/ingest system
         felis: `dict`, optional
             Felis schema for tables. Defaults to None.
+
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: extra empty line is not needed. Please check everything below too.

File: rootfs/ingest/python/qserv/ingest.py [hunk 16]
```diff
@@ -251,24 +250,28 @@ def _ingest_transaction(self) -> bool:
...
+            self.queue_manager.unlock_contribfiles(ingest_success)
         continue_ingest = True
         return continue_ingest
 
     def _ingest_all_contributions(self, transaction_id: int, contributions: list[Contribution]) -> bool:
-        """Ingest all contribution for a given transaction
-           Throw exception if ingest fail
-           This method always returns True, or raise an exception
+        """Ingest all contribution for a given transaction Throw exception if
+        ingest fail This method always returns True, or raise an exception.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: Add periods after "transaction" and "fail".

fail -> fails.
raise -> raises.

File: rootfs/ingest/python/qserv/ingest.py [hunk 17]
```diff
@@ -305,21 +308,20 @@ def _ingest_all_contributions(self, transaction_id: int, contributions: list[Con
                 loop = False
             else:
                 _LOG.info(
-                    "Contributions for transaction %s, RECENTLY STARTED: %s, NOT FINISHED: %s," +
-                    "RECENTLY FINISHED: %s, FINISHED: %s",
+                    "Contributions for transaction %s, RECENTLY STARTED: %s, NOT FINISHED: %s,"
+                    + "RECENTLY FINISHED: %s, FINISHED: %s",
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/ingest.py]: You don't need `+` to concatenate literal strings.
Add space before `RECENTLY`.

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 18]
```diff
@@ -77,18 +77,20 @@ class TransactionState(Enum):
 
 
 class ContributionMonitor:
-    """Store contribution status returned by the Ingest Service
-    see https://confluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol
-    for details
+    """Store contribution status returned by the Ingest Service see https://con
+    fluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol for
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py]: Line break in the middle of URL

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 19]
```diff
@@ -77,18 +77,20 @@ class TransactionState(Enum):
...
-        response_json: `dict`
-            Ingest Service response for a HTTP request against URL: ingest/file-async/<request_id>
+    response_json: `dict`
+        Ingest Service response for a HTTP request
+        against URL: ingest/file-async/<request_id>
 
     Raises
     ------
-        ReplicationControllerError: If 'response_json' does not have the expected value or format
+    error: `ReplicationControllerError`
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py]: Just `ReplicationControllerError`.

Please review the style guide https://developer.lsst.io/python/numpydoc.html and try to follow that.

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 20]
```diff
@@ -195,20 +198,23 @@ def get_fqdn(fqdns: str, port: int, scheme: str = "http") -> str:
...
     ----------
     responseJson: `dict`
         Response for replication service for the regular table locations API
-        see https://confluence.lsstcorp.org/pages/viewpage.action?pageId=133333850#UserguidefortheQservIngestsystem(APIversion8)-Locateregulartables
+        see
+        https://confluence.lsstcorp.org/pages/viewpage.action?pageId=133333850#UserguidefortheQservIngestsystem(APIversion8)-Locateregulartables
 
     Returns
     -------
     locations `typing.List[typing.Tuple[str, int]]`:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py:213]: Colon must be after `locations`.

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 21]
```diff
@@ -195,20 +198,23 @@ def get_fqdn(fqdns: str, port: int, scheme: str = "http") -> str:
...
-        List of connection parameters , for Data ingest Service REST API i.e. http,
-        of workers which are available for ingesting regular (fully replicated) tables
+        List of connection parameters for Data ingest Service REST API,
+        for workers which are available for ingesting
+        regular (fully replicated) tables
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py]: This formatting looks odd.

File: rootfs/ingest/python/qserv/jsonparser.py [hunk 22]
```diff
@@ -279,8 +287,10 @@ def raise_error(responseJson: dict, retry_attempts: int = -1, max_retry_attempts
     Returns
     -------
     is_error_retryable: `bool`
-        True if retry_attempts < max_retry_attempts and if error allows retrying request
+        True if retry_attempts < max_retry_attempts
+        and if error allows retrying request
         False if no error in JSON response
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/jsonparser.py]: Needs period after `request`. Formatting looks odd here too.

File: rootfs/ingest/python/qserv/loadbalancerurl.py [hunk 23]
```diff
@@ -64,15 +64,15 @@ def get(self) -> Optional[str]:
 
 
 class LoadBalancedURL:
-    """Manage http(s) load balanced URL
-    Also file file:// protocol, and use it as default if no scheme is provided
-    """
+    """Manage http(s) load balanced URL Also file file:// protocol, and use it
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/loadbalancerurl.py]: Missing period.

File: rootfs/ingest/python/qserv/queue.py [hunk 24]
```diff
@@ -62,17 +62,15 @@
 
 
 class QueueManager:
-    """
-    Class implementing contributions queue manager for Qserv ingest process
-    """
+    """Class implementing contributions queue manager for Qserv ingest
+    process."""
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py]: Closing `"""` should be on a separate line.

File: rootfs/ingest/python/qserv/queue.py [hunk 25]
```diff
@@ -217,27 +213,23 @@ def _acquire_mutex(self) -> None:
...
         release_mutex_query = release_mutex_query.where(self.mutex.c.pod == self.pod)
         self._safe_execute(release_mutex_query, _MAX_RETRY_ATTEMPTS)
 
     def init_mutex(self) -> None:
-        """Initialize mutex in queue database
-           Queue database has a table `mutex` which contain only one row. This row is used to enable
-           only one pod to lock contribution file in queue at a time and need to be initialized
-           at ingest startup.
+        """Initialize mutex in queue database Queue database has a table
+        `mutex` which contain only one row.
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py]: Needs double backticks around `mutex` here, as it does not refer to a type or variable.

File: rootfs/ingest/python/qserv/queue.py [hunk 26]
```diff
@@ -280,35 +269,43 @@ def _run_lock_queries(self, contribfiles_to_lock_count: int) -> int:
...
         _LOG.debug("%s contribution files locked by pod %s", contribfiles_locked_count, self.pod)
 
         return contribfiles_locked
 
     def unlock_contribfiles(self, ingest_success: bool) -> None:
-        """Mark contributions as "succeed" in contribution queue if super-transaction
-        has been successfully commited
-        Release contributions in queue when the super-transaction has been aborted
+        """Mark contributions as "succeed" in contribution queue if super-
+        transaction has been successfully commited Release contributions in
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py]: Missing period

File: rootfs/ingest/python/qserv/queue.py [hunk 27]
```diff
@@ -376,9 +375,7 @@ def _safe_execute(self, query: typing.Any, max_retry: int = 0) -> None:
                 if retry_count < max_retry:
                     _LOG.error(
                         "Database connection error: {} - sleeping for {}s"
-                        " and will retry (attempt #{} of {})".format(
-                            ex, wait_sec, retry_count, max_retry
-                        )
+                        " and will retry (attempt #{} of {})".format(ex, wait_sec, retry_count, max_retry)
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/queue.py]: Do not use `{}` and `format` with logging, replace with `%s`

File: rootfs/ingest/python/qserv/replicationclient.py [hunk 28]
```diff
@@ -111,24 +104,26 @@ def close_transaction(self, database: str, transaction_id: int, success: bool) -
...
     def _check_version(self) -> None:
         url = urllib.parse.urljoin(self.repl_url, "meta/version")
-        responseJson = self.http.get(url, timeout=self.timeout_short)
+        responseJson = self.http.get(url, timeout=TIMEOUT_SHORT_SEC)
         jsonparser.raise_error(responseJson)
         if responseJson["version"] != _REPL_SERVICE_VERSION:
             raise ValueError(
-                "Invalid replication server version " + f"(is {responseJson['version']}, expected {_REPL_SERVICE_VERSION})"
+                "Invalid replication server version "
+                + f"(is {responseJson['version']}, expected {_REPL_SERVICE_VERSION})"
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/replicationclient.py]: `+` is not needed

File: rootfs/ingest/python/qserv/replicationclient.py [hunk 29]
```diff
@@ -227,7 +217,7 @@ def get_chunk_location(self, chunk_id: int, database: str) -> Tuple[str, int]:
     @lru_cache(maxsize=None)
     def get_regular_tables_locations(self, database: str) -> List[Tuple[str, int]]:
         """Returns connection parameters of the Data Ingest Service of workers
-           which are available for ingesting regular (fully replicated) tables:
+        which are available for ingesting regular (fully replicated) tables:
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/replicationclient.py]: colon should be aperiod?

File: rootfs/ingest/python/qserv/test_http.py [hunk 30]
```diff
@@ -61,18 +61,23 @@ def test_json_get() -> None:
...
     with pytest.raises(HTTPError) as e:
         _http.get(url="https://www.in2p3.cnrs.fr/notfound", payload={}, auth=False)
     assert e.value.args[0] == "404 Client Error: Not Found for url: https://www.in2p3.cnrs.fr/notfound"
 
 
 def test_retry() -> None:
-    """Check if a retry occurs for a non-existing DNS entry
-    This might occurs if k8s DNS fails intermittently
-    """
+    """Check if a retry occurs for a non-existing DNS entry This might occurs
```
Comment (andy-slac) [inline rootfs/ingest/python/qserv/test_http.py]: missing period

**Review by andy-slac — APPROVED** [review]
Looks OK in general, but there are inconsistencies in docstring formatting and docstring/comment wrapping. Please review the LSST style guide and try to follow that.

File: codeformat.sh (hunk 1 above)
Comment (fjammes) [reply codeformat.sh:8]: :+1:

File: rootfs/ingest/python/qserv/contribution.py (hunk 4 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/contribution.py]: :+1:

File: doc/conf.py (hunk 2 above)
Comment (fjammes) [reply doc/conf.py]: :+1:

File: rootfs/ingest/python/qserv/http.py (hunk 9 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/http.py]: @andy-slac writing the correct docstring format is complicated and time-consuming. I do not understand why LSST does not use a tool to automatically generate dosctring, that the developer could then fill with useful information. I can easilly do that in VS code plugin (by typing """ below the function declaration), but unfortunately the docstring does not fit the LSST standards:
```
    def myfunc(
        self, url: str, payload: Dict[str, Any] = dict(), auth: bool = True, timeout: Optional[int] = None
    ) -> Dict:
        """_summary_

        Args:
            url (str): _description_
            payload (Dict[str, Any], optional): _description_. Defaults to dict().
            auth (bool, optional): _description_. Defaults to True.
            timeout (Optional[int], optional): _description_. Defaults to None.

        Returns:
            Dict: _description_
        """
```
Is there such a code editor plugin which generate docstring according to the LSST standards?

Comment (andy-slac) [reply rootfs/ingest/python/qserv/http.py]: @fjammes, reviewing a pull request is time consuming. Please try to make sure that your code follows LSST conventions when you send it for review.

**Pull request merged**